group-coordinator/src/main/java/org/apache/kafka/coordinator/group/consumer/CurrentAssignmentBuilder.java

```diff
@@ -34,48 +33,68 @@
  * consumer group protocol. Given the current state of a member and a desired or target
  * assignment state, the state machine takes the necessary steps to converge them.
  *
- * The member state has the following properties:
- * - Current Epoch:
- *   The current epoch of the member.
+ * State Machine:
+ * - STABLE:
+ *   The member is fully reconciled to the desired target assignment.
  *
- * - Next Epoch:
- *   The desired epoch of the member. It corresponds to the epoch of the target/desired assignment.
- *   The member transitions to this epoch when it has revoked the partitions that it does not own
- *   or if it does not have to revoke any.
+ *   Valid Transitions:
+ *   - STABLE -> STABLE
+ *     When a new target assignment is installed but the assignment of the member
+ *     has not changed, the member transitions to the next epoch and remains in
+ *     the STABLE state.
  *
- * - Previous Epoch:
- *   The epoch of the member when the state was last updated.
+ *   - STABLE -> UNACKNOWLEDGED_ASSIGNMENT
+ *     When a new target assignment is installed and a new assignment is computed
+ *     for the member, he transitions to the UNACKNOWLEDGED_ASSIGNMENT state.
  *
- * - Assigned Partitions:
- *   The set of partitions currently assigned to the member. This represents what the member should have.
+ *     If the next assignment contains partitions to be revoked, the member stays
+ *     in his current epoch. Otherwise, he transitions to the target epoch.
  *
- * - Partitions Pending Revocation:
- *   The set of partitions that the member should revoke before it can transition to the next state.
+ *   - STABLE -> UNRELEASED_PARTITIONS
```

Comment:
For this transition, I'm confused which one is the default first transition. Ie, if the client has not yet acknowledged AND we have unreleased partitions which one do we go to first? And is going between these two both ways only possible if a new assignment occurs in the middle? I'm struggling to understand how all transitions are valid 😅

Reply:
I'm confused about this as well, I would think it would go from stable -> unack (when new assignment computed) -> unreleased (if there's unreleased partitions). If it's possible to ack the assignment even when there's unreleased partitions, we're transitioning from unack to unreleased right?